pkg/validation/options.go

```diff
@@ -29,28 +29,40 @@ func Validate(o *options.Options) error {
 	msgs = configureLogger(o.Logging, msgs)
 	msgs = parseSignatureKey(o, msgs)
 
+	var tlsConfig *tls.Config
 	if o.SSLInsecureSkipVerify {
 		// InsecureSkipVerify is a configurable option we allow
 		/* #nosec G402 */
-		insecureTransport := &http.Transport{
-			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		tlsConfig = &tls.Config{
+			InsecureSkipVerify: true,
 		}
-		http.DefaultClient = &http.Client{Transport: insecureTransport}
 	} else if len(o.Providers[0].CAFiles) > 0 {
 		pool, err := util.GetCertPool(o.Providers[0].CAFiles)
 		if err == nil {
-			transport := http.DefaultTransport.(*http.Transport).Clone()
-			transport.TLSClientConfig = &tls.Config{
+			tlsConfig = &tls.Config{
 				RootCAs:    pool,
 				MinVersion: tls.VersionTLS12,
 			}
-
-			http.DefaultClient = &http.Client{Transport: transport}
 		} else {
 			msgs = append(msgs, fmt.Sprintf("unable to load provider CA file(s): %v", err))
 		}
 	}
 
+	if len(o.Providers[0].CertFiles) > 0 && len(o.Providers[0].KeyFiles) > 0 {
+		certs, err := util.GetClientCertificates(o.Providers[0].CertFiles, o.Providers[0].KeyFiles)
+		if err == nil {
+			tlsConfig.Certificates = certs
+		} else {
+			msgs = append(msgs, fmt.Sprintf("unable to load client certificates: %v", err))
+		}
+	}
+
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.TLSClientConfig = tlsConfig
+	http.DefaultClient = &http.Client{
+		Transport: transport,
+	}
```

Comment:
Pushing the config to the providers makes sense. The only place I get a bit tripped up in doing that is places like this https://github.com/oauth2-proxy/oauth2-proxy/blob/39ef5de97033fdb4b0253223c253ddb0bd1c9123/pkg/providers/oidc/provider.go#L58. May just be specific to my use case, but the oidc endpoints are behind mTLS as well. So aside from passing clients through the context it turns into a bigger update. The default client was just easier to override because of that, with minimal changes to the existing config structure. In my case, I just need to be able to add tls configs to talk to the backend auth services, so if replacing the default client is too invasive I'm sure I can rework a few things.